PowerPointLabs/PowerPointLabs/CropLab/CropToAspectRatio.cs

```diff
@@ -0,0 +1,224 @@
+        {
+            string pattern = @"(\d+):(\d+)";
+            Match matches = Regex.Match(aspectRatioString, pattern);
+            if (!matches.Success)
+            {
+                ThrowErrorCode(ErrorCodeForAspectRatioInvalid);
+            }
+
+            if (!float.TryParse(matches.Groups[1].Value, out aspectRatioWidth) ||
+                !float.TryParse(matches.Groups[2].Value, out aspectRatioHeight))
+            {
+                ThrowErrorCode(ErrorCodeForAspectRatioInvalid);
+            }
+            
+            if (aspectRatioWidth <= 0.0f || aspectRatioHeight <= 0.0f)
+            {
+                ThrowErrorCode(ErrorCodeForAspectRatioInvalid);
+            }
+        }
+
+        private static void VerifyIsSelectionValid(PowerPoint.Selection selection)
+        {
+            if (selection.Type != PowerPoint.PpSelectionType.ppSelectionShapes)
+            {
+                ThrowErrorCode(ErrorCodeForSelectionCountZero);
+            }
+        }
+
+        private static bool IsPictureForSelection(PowerPoint.ShapeRange shapeRange)
+        {
+            return (from PowerPoint.Shape shape in shapeRange select shape).All(IsPicture);
+        }
+
+        private static bool IsPicture(PowerPoint.Shape shape)
+        {
+            return shape.Type == Office.MsoShapeType.msoPicture;
+        }
+
+        private static void ThrowErrorCode(int typeOfError)
+        {
+            throw new Exception(typeOfError.ToString(CultureInfo.InvariantCulture));
+        }
+
+        private static void IgnoreExceptionThrown(){}
+
+        public static string GetErrorMessageForErrorCode(string errorCode)
+        {
+            var errorCodeInteger = -1;
+            try
+            {
+                errorCodeInteger = Int32.Parse(errorCode);
+            }
+            catch
+            {
+                IgnoreExceptionThrown();
+            }
+            switch (errorCodeInteger)
+            {
+                case ErrorCodeForSelectionCountZero:
+                    return ErrorMessageForSelectionCountZero;
+                case ErrorCodeForSelectionNonPicture:
+                    return ErrorMessageForSelectionNonPicture;
+                case ErrorCodeForAspectRatioInvalid:
+                    return ErrorMessageForAspectRatioInvalid;
+                default:
+                    return ErrorMessageForUndefined;
+            }
+        }
+
+        private static void ProcessErrorMessage(Exception e)
+        {
+            //This method prompts the error message to user. If it has an unrecognised error code,
+            //an alternative message window with erro trace stack pops up and prompts the user to
+            //send the trace stack to the developer team.
+            var errMessage = GetErrorMessageForErrorCode(e.Message);
+            if (!string.Equals(errMessage, ErrorMessageForUndefined, StringComparison.Ordinal))
+            {
+                MessageBox.Show(errMessage, MessageBoxTitle);
```

Comment:
find a way to wrap it, so that the codes are testable in UT (UT cannot handle a msg box)